src/c_kzg_4844.c

```diff
@@ -2611,102 +2465,68 @@ static C_KZG_RET zero_polynomial_via_multiplication(
-        partials[partial_count - 1].length = 1 + len_missing -
-                                             (partial_count - 1) *
-                                                 missing_per_partial;
-
-        /* Reduce all the partials to a single polynomial */
-        while (partial_count > 1) {
-            size_t reduced_count = (partial_count + reduction_factor - 1) /
-                                   reduction_factor;
-            size_t partial_size = next_power_of_two(partials[0].length);
-            for (size_t i = 0; i < reduced_count; i++) {
-                size_t start = i * reduction_factor;
-                size_t out_end = MIN(
-                    (start + reduction_factor) * partial_size, n
-                );
-                size_t reduced_len = MIN(
-                    out_end - start * partial_size, s->max_width
-                );
-                size_t partials_num = MIN(
-                    reduction_factor, partial_count - start
-                );
-                partials[i].coeffs = work + start * partial_size;
-                if (partials_num > 1) {
-                    ret = reduce_partials(
-                        &partials[i],
-                        reduced_len,
-                        scratch,
-                        n * 3,
-                        &partials[start],
-                        partials_num,
-                        s
-                    );
-                    if (ret != C_KZG_OK) goto out;
-                } else {
-                    partials[i].length = partials[start].length;
-                }
-            }
-            partial_count = reduced_count;
-        }
+        In other words, if the missing index is `i`, then we add \omega^i
+        to the roots array, where \omega is a primitive `CELLS_PER_EXT_BLOB`
+        root of unity.
+    */
+    size_t stride = s->max_width / CELLS_PER_EXT_BLOB;
+    for (size_t i = 0; i < len_missing_cells; i++) {
+        roots[i] = s->expanded_roots_of_unity[missing_cell_indices[i] * stride];
+    }
 
-        /* Pad the output with zeros */
-        ret = pad_p(
-            zero_poly, s->max_width, partials[0].coeffs, partials[0].length
-        );
-        if (ret != C_KZG_OK) goto out;
+    /* Compute the polynomial that evaluates to zero on the roots*/
+    ret = zero_polynomial_from_roots(
+        short_zero_poly, &short_zero_poly_len, roots, len_missing_cells
+    );
+    if (ret != C_KZG_OK) goto out;
 
-        *zero_poly_len = partials[0].length;
-    }
+    /*
+        For each root \omega^i in `short_zero_poly`, we compute a polynomial
+        that has roots at H =   {
+                                \omega^i * \gamma^0,
+                                \omega^i * \gamma^1,
+                                ...,
+                                \omega^i * \gamma^{FIELD_ELEMENTS_PER_CELL-1}
+                                }
+        where \gamma is primitive `FIELD_ELEMENTS_PER_EXT_BLOB` generator.
```

Comment:
I don't fully get what `\gamma is primitive FIELD_ELEMENTS_PER_EXT_BLOB generator` is supposed to mean.

Reply:
It might be the phrasing: It means that \gamma is able to generate the subgroup of size FIELD_ELEMENTS_PER_EXT_BLOB where each element is a root of unity